### PR TITLE
feat(presets): add support to presets ending with `.json5` or `.json`

### LINKED ISSUE #15370: `default.json5` filename for Shareable Config Presets
### What would you like Renovate to be able to do?

Since JSON5 support was added for shared configs in #15187 (original issue #7674), it would be helpful for tooling (eg. formatting tools, syntax highlighting in editors and on GitHub itself, other tooling) to be able to recognize this file as using the JSON5 syntax.

One way of communicating that to tooling is to use the `.json5` file extension.

It would be nice to have the ability of using `default.json5` as [a config presets filename](https://docs.renovatebot.com/config-presets/) (maybe also `renovate.json5`?

### If you have any ideas on how this should be implemented, please tell us here.

It would be best if it was not done with an additional API call, as @rarkins mentions in https://github.com/renovatebot/renovate/issues/7674#issuecomment-1113113323:

> The answer is no if it means increasing the number of API try/catch/try API requests we do. AFAIK we today already have to try default.json and if it's missing then renovate.json, and that's something we want to deprecate already. So adding an additional check for default.json5 would not be great.
>
> Instead, if we can extend our syntax so that you can explicitly specify the filename, then that would be fine. That way there's only one API call. Best to discuss in a new issue, if none already exist (I know we've discussed it before but not sure if we have a feature request open)

So it could be possibly implemented as a way of configuring the extension in the filename.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by oscard0m**
## Changes
Add support to Renovate configuration `presets` ending with `.json5` or `.json`

- [x] Add support to presets ending with `.json5`
- [x] Add support to presets ending with `.json`
- [x] Add tests for covering the rest of Preset Sources
- [x] Update documentation

## Context
~I want to validate first if a preset using `.json` extension is treated as expected as pointed by @rarkins in https://github.com/renovatebot/renovate/issues/15370#issuecomment-1113166086~

~EDIT: Support to `.json` presets won't be addressed in the context of this issue (details [here](https://github.com/renovatebot/renovate/issues/15370#issuecomment-1113524382))~

EDIT2: In the end, this PR will tackle the support presets ending with `.json` or `.json5` (details [here](https://github.com/renovatebot/renovate/pull/15377#discussion_r861993446))

Closes #15370

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository [[repo here](https://github.com/oscard0m/test-renovate-repo/pulls)]

### CODE REVIEW

**Comment by rarkins** [comment]
I didn't expect this to work yet. My advice was to continue treating .json as the implicit extension and add new logic specifically for .json5 extensions

File: lib/config/presets/github/index.spec.ts [hunk 1]
```diff
@@ -98,6 +98,20 @@ describe('config/presets/github/index', () => {
...
+        .reply(200, {
+          content: toBase64('{"foo":"bar"}'),
+        });
+      const content = await github.getPreset({
+        repo: 'some/repo',
+        presetName: 'somefile.json5',
+      });
+      expect(content).toEqual({ foo: 'bar' });
+    });
+
```
Comment (oscard0m) [inline lib/config/presets/github/index.spec.ts:114]: It's a change in `presets/utils.ts` but the existing tests are mocking the response of the `fetch` so I think it gives more value in `presets/github/index.spec.ts`

If you agree I can replicate the same tests for the other `presetSources`.

File: lib/config/presets/util.ts [hunk 2]
```diff
@@ -51,7 +51,9 @@ export async function fetchPreset({
   } else {
     jsonContent = await fetch(
       repo,
-      buildFilePath(`${fileName}.json`),
+      buildFilePath(
+        fileName.endsWith('.json5') ? fileName : `${fileName}.json`
```
Comment (viceice) [inline lib/config/presets/util.ts:55]: i like to also support explicit json extension in this PR, should be easy to implement with a regex.

**Review by viceice — COMMENTED** [review]
I'm not sure if we also need to update the regex for preset type detections.

can you please do a real run?

File: lib/config/presets/github/index.spec.ts (hunk 1 above)
Comment (oscard0m) [reply lib/config/presets/github/index.spec.ts:128]: Already pushed for other sources. Let me know if it's ok and I can refactor it if necessary.

**Comment by oscard0m** [comment]
> I'm not sure if we also need to update the regex for preset type detections.
> 
> can you please do a real run?

I've been able to run it without modifying regexes. I think regexes are working for our use case. I still want to run more tests in my local later to double-check this.

Here's the test repo: https://github.com/oscard0m/test-renovate-repo/pulls

**Comment by viceice** [comment]
> > I'm not sure if we also need to update the regex for preset type detections.
> > 
> > can you please do a real run?
> 
> I've been able to run it without modifying regexes. I think regexes are working for our use case. I still want to run more tests in my local later to double-check this because I'm getting re-opened the issue of renovate complaining.
> 
> Here's the test repo: https://github.com/oscard0m/test-renovate-repo/pulls

sounds good, use re-request review button when you need a new review

https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

File: lib/config/presets/util.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ export async function fetchPreset({
   } else {
     jsonContent = await fetch(
       repo,
-      buildFilePath(`${fileName}.json`),
+      buildFilePath(/\.json5?$/.test(fileName) ? fileName : `${fileName}.json`),
```
Comment (viceice) [inline lib/config/presets/util.ts:54]:
```suggestion
      buildFilePath(regEx(/\.json5?$/).test(fileName) ? fileName : `${fileName}.json`),
```
please use our re2 wrapper.

File: lib/config/presets/bitbucket-server/index.spec.ts [hunk 4]
```diff
@@ -35,6 +35,24 @@ describe('config/presets/bitbucket-server/index', () => {
...
     });
 
+    it('returns JSON5', async () => {
+      httpMock
+        .scope(bitbucketApiHost)
+        .get(`${basePath}/some-filename.json5`)
+        .query({ limit: 20000 })
+        .reply(200, {
+          isLastPage: true,
+          lines: [{ text: '{"from":"api"' }, { text: '}' }],
```
Comment (viceice) [inline lib/config/presets/bitbucket-server/index.spec.ts:45]: you should use some JSON5 syntax in those tests. 😉

Comment (oscard0m) [reply lib/config/presets/bitbucket-server/index.spec.ts:45]: > you should use some JSON5 syntax in those tests. 😉

good point! Thanks

File: docs/usage/config-presets.md [hunk 5]
```diff
@@ -55,6 +55,7 @@ You can set a Git tag (like a SemVer) to use a specific release of your shared c
 | ------------------------------------------- | -------------------------------- | --------- | ---------------------------- | --------------- | -------------- |
 | GitHub default                              | `github>abc/foo`                 | `default` | `https://github.com/abc/foo` | `default.json`  | Default branch |
```
Comment (viceice) [inline docs/usage/config-presets.md:56]: @HonkingGoose Can you open a new docs issue to refactor this file? I think it's very verbose to duplicate every possible preset options for every platform.

I like to have one general table and then below platform specific values if exist.

**Review by viceice — DISMISSED** [review]

**Comment by oscard0m** [comment]
> I'm not sure if we also need to update the regex for preset type detections.
> 
> can you please do a real run?

Even though the execution seems to work, I'm getting this issue open. Any clue what could be the issue?

https://github.com/oscard0m/test-renovate-repo/issues/2

File: lib/config/presets/util.ts (hunk 3 above)
Comment (oscard0m) [reply lib/config/presets/util.ts]: 💡  would be nice to lint this (not sure if it exists an ESLint rule for this)

**Comment by rarkins** [comment]
@oscard0m that's the "real" production bot raising the error. i.e. it's running current code and doesn't have your PR logic yet. You'd need to run it locally e.g. using `yarn start ....` from your branch.

**Comment by oscard0m** [comment]
> 

I run `yarn start` in my local machine. I noticed I had `Renovate` App installed in all my repositories so I assume that's why. Already uninstalled in that repo and re-running `yarn start`.

For my learning and understanding: Is there an easy way to see if a PR or an issue is created by a local Renovate (`yarn start ...`) or by a production one @rarkins?

**Review by viceice — APPROVED** [review]

**Pull request merged**